crates/bevy_render/src/mesh/morph.rs

```diff
@@ -0,0 +1,226 @@
+    pub fn new(
+        targets: impl ExactSizeIterator<Item = impl Iterator<Item = MorphAttributes>>,
+        vertex_count: usize,
+    ) -> Result<Self, MorphBuildError> {
+        let max = MAX_TEXTURE_WIDTH;
+        let target_count = targets.len();
+        if target_count > MAX_MORPH_WEIGHTS {
+            return Err(MorphBuildError::TooManyTargets { target_count });
+        }
+        let component_count = (vertex_count * MorphAttributes::COMPONENT_COUNT) as u32;
+        let Some((Rect(width, height), padding)) = lowest_2d(component_count , max) else {
+            return Err(MorphBuildError::TooManyAttributes { vertex_count, component_count });
+        };
+        let data = targets
+            .flat_map(|mut attributes| {
+                let layer_byte_count = (padding + component_count) as usize * mem::size_of::<f32>();
+                let mut buffer = Vec::with_capacity(layer_byte_count);
+                for _ in 0..vertex_count {
+                    let Some(to_add) = attributes.next() else {
+                        break;
+                    };
+                    buffer.extend_from_slice(bytemuck::bytes_of(&to_add));
+                }
+                // Pad each layer so that they fit width * height
+                buffer.extend(iter::repeat(0).take(padding as usize * mem::size_of::<f32>()));
+                debug_assert_eq!(buffer.len(), layer_byte_count);
+                buffer
+            })
+            .collect();
+        let extents = Extent3d {
+            width,
+            height,
+            depth_or_array_layers: target_count as u32,
+        };
+        let image = Image::new(extents, TextureDimension::D3, data, TextureFormat::R32Float);
+        Ok(MorphTargetImage(image))
+    }
+}
+
+/// Control a [`Mesh`]'s [morph targets].
+///
+/// Add this to an [`Entity`] with a [`Handle<Mesh>`] with a [`MorphAttributes`] set
+/// to control individual weights of each morph target.
+///
+/// [morph targets]: https://en.wikipedia.org/wiki/Morph_target_animation
+/// [`Entity`]: bevy_ecs::prelude::Entity
+#[derive(Reflect, Default, Debug, Clone, Component)]
+#[reflect(Debug, Component)]
+pub struct MorphWeights {
+    weights: Vec<f32>,
+}
+impl MorphWeights {
+    pub fn new(weights: Vec<f32>) -> Result<Self, MorphBuildError> {
+        if weights.len() > MAX_MORPH_WEIGHTS {
+            let target_count = weights.len();
+            return Err(MorphBuildError::TooManyTargets { target_count });
+        }
+        Ok(MorphWeights { weights })
+    }
+    pub fn weights(&self) -> &[f32] {
+        &self.weights
+    }
+    pub fn weights_mut(&mut self) -> &mut [f32] {
+        &mut self.weights
+    }
+}
+
+/// Bevy meshes are gltf primitives, [`MorphWeights`] on the bevy node entity
+/// should be inherited by children meshes.
+///
+/// Only direct children are updated, to fulfill the expectations of glTF spec.
+pub fn inherit_weights(
```

Comment:
"Weight inheritance" feels a bit gnarly to me. It makes using this system hard because you just have to "know" that a given `MorphWeights` references is canonical ("top level" node) or mirrored ("mesh primitve node"). I'm thinking we should treat each "primitive" as fully separate / with its own weights. Bevy has no concept of a "higher level" organizational mesh like GLTF does and I'd like "bevy apis" to avoid embracing that pattern wherever possible.

I _think_ this is a better path (but I'm curious what everyone else thinks):
1. Remove the "top level" MorphWeights
2. When loading the GLTF, if there is an animation that points to a node with multiple primitives, duplicate it across those primitives (ex: call `add_curve_to_path` once for each `base_node/Primitive0` path).

This ensures animations still play "together" (as intended by the gltf), but it allows us to treat them as fully separate things when necessary (from a bevy perspective) and leaves MorphWeights as a simpler, more general 1:1 mesh abstraction (assumed to always be paired with a `Handle<Mesh>`). Devs can modify them directly without worrying about "writing the wrong one and getting overwritten" and they don't need to play the `With<Handle<Mesh>>` / `Without<Handle<Mesh>>` games currently required.

Reply:
Or alternatively we could embrace the "multiple surfaces in a single Mesh" model via something like Godot's [ArrayMesh](https://docs.godotengine.org/en/stable/classes/class_arraymesh.html) (but thats the kind of thing we should probably punt). Then we wouldn't have this problem because there would only be one entity containing multiple mesh "surfaces".

Probably worth doing a survey of the space across engines. I personally really like the simplicity of the "one mesh / material per entity" model as that is how things are actually rendered (and it makes it easy for users to think about things). But if enough people expect this functionality we should probably accommodate that.

Supporting ArrayMesh would also probably require a tighter coupling between Mesh and Material, which would complicate a number of existing systems / force reworks.

Reply:
What about splitting `MorphWeights` in two? A publicly exposed `MorphWeights` that the user can access and modify, and a private `MeshMorphWeights` managed by the engine. (or similarly to `Transform`/`GlobalTransform`, both could be public and only one is publicly mutable) The public one would be the "root" controller while the private ones are the mesh components.

This would solve the footgun of the `With/Without`.

I'm worried with the proposal of splitting the `MorphWeights` and using `add_curve_to_path`. Wouldn't that break backed-in animations? As a user, I would be surprised to see my file with 3 animations suddenly loaded with 7 ones because the morph one was split. If the user has a named animation, then it's unclear how we'd handle them. Should we copy the name? In engine code, it also requires identifying the animations touching morph targets and treat them specially.

Reply:
> What about splitting MorphWeights in two? A publicly exposed MorphWeights that the user can access and modify, and a private MeshMorphWeights managed by the engine. (or similarly to Transform/GlobalTransform, both could be public and only one is publicly mutable) The public one would be the "root" controller while the private ones are the mesh components.

This is definitely better, as it solves some of the usability problems. And it _might_ be the right path forward for 0.11 (as we're running out of time here). But imo it still feels suboptimal from a "user mental model" perspective in the context of Bevy. Feels too reliant on "opaque arbitrary rules about entity nesting and structure", when this is a "mesh feature".

> I'm worried with the proposal of splitting the MorphWeights and using add_curve_to_path. Wouldn't that break backed-in animations?

Yeah I don't love the duplication. But I don't really see how it would break baked in animations. They would all still play at the same time?

> As a user, I would be surprised to see my file with 3 animations suddenly loaded with 7 ones because the morph one was split. If the user has a named animation, then it's unclear how we'd handle them. Should we copy the name? In engine code, it also requires identifying the animations touching morph targets and treat them specially.

Note that it doesn't create new _animations_ (as in AnimationClips). It just adds splits out curves within the top level clip.

For example this:
```
- AnimationClip: "Smile"
  - Curve: "Root/MyGltfMesh" : Value1, Value2, Value3
``` 

Would become this:
```
- AnimationClip: "Smile"
  - Curve: "Root/MyGltfMesh/FacePrimitive" : Value1, Value2, Value3
  - Curve: "Root/MyGltfMesh/LipPrimitive" : Value1, Value2, Value3
```

Its essentially doing the "mirroring" work we're currently doing, just baked directly into the clip.

Reply:
I've made the changes above. I personally still think they're the right call unless the "mesh surface/primitive situation" changes. We _should_ discuss that, as it would resolve stuff like this and most engines have an equivalent concept.

Reply:
We've opted to go for "inheritance with component split" (see conversations below).